src/Loops.purs

```diff
@@ -1,149 +1,171 @@
 module Loops
-  ( Loop
+  ( Sample (..)
+  , Passage
+  , merge
+  , silence
   , bd
   , sn
   , hh
-  , silence
-  , BPM
-  , Audio
+  , Millis
+  , Dur (..)
+  , Event
   , Track
+  , unsafeWarp
   , loop
+  , Audio
+  , Playable
   , track
+  , play
+  , degrade
   ) where
 
 import Prelude
 import Data.List.Lazy as Lazy
-import Audio.Howler (HOWLER, defaultProps, new, play)
+import Audio.Howler (HOWLER, defaultProps, new, play) as H
+import Data.Rational (Rational, (%), toNumber, fromInt)
+import Data.Int (round)
+import Data.Int (toNumber) as Int
 import Control.Monad.Eff (Eff)
 import Control.Monad.Eff.Ref (REF, newRef, readRef, writeRef)
-import Control.Monad.Eff.Timer (TIMER, clearInterval, setInterval, setTimeout)
+import Control.Monad.Eff.Timer (TIMER, setInterval, setTimeout)
+import Data.Newtype (wrap, class Newtype, over, over2)
 import Data.Foldable (for_)
-import Data.List (List, singleton)
+import Data.List (List(..), (:), singleton)
 import Data.Monoid (class Monoid, mempty)
-import Data.Newtype (class Newtype, unwrap)
+
+import Control.Monad.Eff.Random (RANDOM, random)
+
+-- | Requires a monotonic increasing function of numbers
+unsafeWarp :: (Number -> Number) -> Track -> Track
+unsafeWarp f = over Track $ map \{sample,time} -> {sample, time: f' time}
+  where f' t = round $ x * 1000.0
+          where x = f $ Int.toNumber t / 1000.0
 
 data Sample = Bd | Sn | Hh
 
--- | A `Loop` is a finite list of timed samples, with a duration.
--- |
--- | `Loop` is a `Semigroup` and a `Monoid` so it is simple to build loops by
--- | concatenating simpler loops:
--- |
--- | ```purescript
--- | bd <> sn :: Loop
--- | ```
--- |
--- | A `Loop` can be turned into a `Track` using the `loop` function.
-newtype Loop = Loop
-  { length :: Int
-  , values :: List { sample :: Sample, offset :: Int }
+newtype Passage = Passage
+  { length :: Rational
+  , values :: List { sample :: Sample, offset :: Rational }
+  }
+
+type PassageUnsized = List { sample :: Sample, offset :: Rational }
+
+shuffleKeepingSort :: PassageUnsized -> PassageUnsized -> PassageUnsized
```

Comment:
Can you please move this into `merge` under a `where` clause and rename it to something shorter?